spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -0,0 +1,531 @@
+	 * <li>MANUAL: Listener is responsible for acking - use a
+	 * {@link AcknowledgingMessageListener}.
+	 * </ul>
+	 * @param ackMode the {@link AckMode}; default BATCH.
+	 */
+	public void setAckMode(AckMode ackMode) {
+		this.ackMode = ackMode;
+	}
+
+	public ErrorHandler getErrorHandler() {
+		return errorHandler;
+	}
+
+	public void setErrorHandler(ErrorHandler errorHandler) {
+		this.errorHandler = errorHandler;
+	}
+
+	public int getConcurrency() {
+		return concurrency;
+	}
+
+	/**
+	 * The maximum number of concurrent {@link MessageListener}s running. Messages from
+	 * within the same partition will be processed sequentially.
+	 * Concurrency must be 1 when explicitly suppplying partitions.
+	 * @param concurrency the concurrency maximum number
+	 */
+	public void setConcurrency(int concurrency) {
+		this.concurrency = concurrency;
+	}
+
+	/**
+	 * The timeout for waiting for each concurrent {@link MessageListener} to finish on
+	 * stopping.
+	 * @param stopTimeout timeout in milliseconds
+	 * @since 1.1
+	 */
+	public void setStopTimeout(int stopTimeout) {
+		this.stopTimeout = stopTimeout;
+	}
+
+	public int getStopTimeout() {
+		return stopTimeout;
+	}
+
+	public Executor getFetchTaskExecutor() {
+		return fetchTaskExecutor;
+	}
+
+	/**
+	 * The task executor for fetch operations.
+	 * @param fetchTaskExecutor the Executor for fetch operations
+	 */
+	public void setFetchTaskExecutor(Executor fetchTaskExecutor) {
+		this.fetchTaskExecutor = fetchTaskExecutor;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return autoStartup;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	@Override
+	public void start() {
+		synchronized (lifecycleMonitor) {
+			if (!running) {
+				if (this.fetchTaskExecutor == null) {
+					this.fetchTaskExecutor = new SimpleAsyncTaskExecutor(
+							(getBeanName() == null ? "" : getBeanName() + "-") + "kafka-fetch-");
+				}
+				if (this.partitions != null && this.concurrency > 1) {
+					logger.warn("When specific partitions are provided, the concurrency must be 1 to avoid "
+							+ "duplicate message delivery; reduced from " + this.concurrency + " to 1");
+					this.concurrency = 1;
+				}
+				for (int i = 0; i < this.concurrency; i++) {
+					this.running = true;
+					FetchTask fetchTask = new FetchTask();
+					this.fetchTasks.add(fetchTask);
+					fetchTaskExecutor.execute(fetchTask);
+				}
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		stop(null);
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		synchronized (lifecycleMonitor) {
+			if (running) {
+				this.running = false;
+				for (FetchTask fetchTask : this.fetchTasks) {
+					fetchTask.stop();
+				}
+			}
+		}
+		if (callback != null) {
+			callback.run();
+		}
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return 0;
+	}
+
+	public class FetchTask implements SchedulingAwareRunnable {
```

Comment:
Wondering how hard would it be to make this a `MessageListenerContainer` of its own and let the surround class be a composite, and only deal with topics. Thus, if someone wanted to create a partition-oriented consumer (hint: SCSt), they would already have a high level single threaded component they can pass a preconfigured Consumer to.